api/src/main/java/io/cloudevents/message/impl/BaseGenericBinaryMessageImpl.java

```diff
@@ -0,0 +1,71 @@
+package io.cloudevents.message.impl;
+
+import io.cloudevents.SpecVersion;
+import io.cloudevents.message.BinaryMessageVisitor;
+import io.cloudevents.message.BinaryMessageVisitorFactory;
+import io.cloudevents.message.MessageVisitException;
+
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+/**
+ * This class implements a BinaryMessage, providing commong logic to most protocol bindings
+ * which supports both Binary and Structured mode.
+ * Content-type is handled separately using a key not prefixed with CloudEvents header prefix.
+ *
+ * @param <HV> Header value type
+ */
+public abstract class BaseGenericBinaryMessageImpl<HV> extends BaseBinaryMessage {
+
+    private final SpecVersion version;
+    private final byte[] body;
+
+    protected BaseGenericBinaryMessageImpl(SpecVersion version, byte[] body) {
+        Objects.requireNonNull(version);
+        this.version = version;
+        this.body = body;
+    }
+
+    @Override
+    public <T extends BinaryMessageVisitor<V>, V> V visit(BinaryMessageVisitorFactory<T, V> visitorFactory) throws MessageVisitException, IllegalStateException {
+        BinaryMessageVisitor<V> visitor = visitorFactory.createBinaryMessageVisitor(this.version);
+
+        // Grab from headers the attributes and extensions
+        this.forEachHeader((k, v) -> {
+            try {
+                if (k.substring(0, getHeaderKeyPrefix().length()).equalsIgnoreCase(getHeaderKeyPrefix())) {
+                    String name = k.substring(getHeaderKeyPrefix().length()).toLowerCase();
+                    if (name.equals("specversion")) {
+                        return;
+                    }
+                    if (this.version.getAllAttributes().contains(name)) {
+                        visitor.setAttribute(name, headerValueToString(v));
+                    } else {
+                        visitor.setExtension(name, headerValueToString(v));
+                    }
+                }
+                if (k.equalsIgnoreCase(getContentTypeHeaderKey())) {
+                    visitor.setAttribute("datacontenttype", headerValueToString(v));
+                }
+            } catch (StringIndexOutOfBoundsException ex) {
+                // String is smaller than 3 characters and it's not equal for sure to CE_PREFIX
+            }
+        });
+
+        // Set the payload
+        if (this.body != null && this.body.length != 0) {
+            visitor.setBody(this.body);
+        }
+
+        return visitor.end();
+    }
+
+    protected abstract String getContentTypeHeaderKey();
```

Comment:
Could use type parameterization for key's type

Reply:
No we can't because in any case the **key** must be string, because during the header processing the method `equalsIgnoreCase` is used.

Reply:
Then you can use `protected abstract boolean isContentTypeHeader(K key)`, especially given that some frameworks already provide case-insensitive headers